Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+        }
+      case .clientOpenServerClosed:
+        // The server has closed, so it makes no sense to send the rest of the request.
+        // Do nothing.
+        ()
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is closed, cannot send a message.")
+      }
+    }
+    
+    mutating func send(status: String, trailingMetadata: Metadata) throws {
+      throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client cannot send status and trailer.")
+    }
+    
+    /// Returns the client's next request to the server.
+    /// - Returns: The request to be made to the server.
+    mutating func nextOutboundMessage() throws -> ByteBuffer? {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is not open yet.")
+      case .clientOpenServerIdle(var clientOpenState):
+        let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+        self.state = .clientOpenServerIdle(clientOpenState)
+        return request
+      case .clientOpenServerOpen(var clientOpenState):
+        let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+        self.state = .clientOpenServerOpen(clientOpenState)
+        return request
+      case .clientOpenServerClosed:
+        // Nothing to do: no point in sending request if server is closed.
+        return nil
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Can't send request if client is closed.")
+      }
+    }
+    
+    mutating func receive(metadata: Metadata, endStream: Bool) throws -> OnMetadataReceived {
+      // This is metadata received by the client from the server.
+      // It can be initial, which confirms that the server is now open;
+      // or an END_STREAM trailer, meaning the response is over.
+      if endStream {
+        try self.clientReceivedEndHeader()
+      } else {
+        try self.clientReceivedMetadata()
+      }
+      return .doNothing
+    }
+    
+    mutating func clientReceivedEndHeader() throws {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't have received a stream end trailer if both client and server are idle.")
+      case .clientOpenServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent an end stream header if it is still idle.")
+      case .clientOpenServerOpen(let state):
+        self.state = .clientOpenServerClosed(.init(previousState: state))
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is already closed, can't have received the end stream trailer twice.")
```

Comment:
It _might_ make sense to ignore this: the server may send headers with 'grpc-status' but without end stream set followed by empty headers with end stream set. We should just ignore it if that happens.